SDL/Quintuplet.cu

```diff
@@ -602,138 +635,394 @@ __device__ bool SDL::passT5RZConstraint(struct SDL::modules& modulesInGPU, struc
     const int layer5 = modulesInGPU.layers[lowerModuleIndex5] + 6 * (modulesInGPU.subdets[lowerModuleIndex5] == SDL::Endcap) + 5 * (modulesInGPU.subdets[lowerModuleIndex5] == SDL::Endcap and modulesInGPU.moduleType[lowerModuleIndex5] == SDL::TwoS);
 
     //slope computed using the internal T3s
-    const int moduleLayer1 = modulesInGPU.moduleType[lowerModuleIndex1];
-    const int moduleLayer2 = modulesInGPU.moduleType[lowerModuleIndex2];
-    const int moduleLayer3 = modulesInGPU.moduleType[lowerModuleIndex3];
-    const int moduleLayer4 = modulesInGPU.moduleType[lowerModuleIndex4];
-    const int moduleLayer5 = modulesInGPU.moduleType[lowerModuleIndex5];
+    const int moduleType1 = modulesInGPU.moduleType[lowerModuleIndex1]; //0 is ps, 1 is 2s
+    const int moduleType2 = modulesInGPU.moduleType[lowerModuleIndex2];
+    const int moduleType3 = modulesInGPU.moduleType[lowerModuleIndex3];
+    const int moduleType4 = modulesInGPU.moduleType[lowerModuleIndex4];
+    const int moduleType5 = modulesInGPU.moduleType[lowerModuleIndex5];
 
-    float slope;
-    if(moduleLayer1 == 0 and moduleLayer2 == 0 and moduleLayer3 == 1) //PSPS2S
+    const float& x1 = mdsInGPU.anchorX[firstMDIndex]/100;
+    const float& x2 = mdsInGPU.anchorX[secondMDIndex]/100;
+    const float& x3 = mdsInGPU.anchorX[thirdMDIndex]/100;
+    const float& x4 = mdsInGPU.anchorX[fourthMDIndex]/100;
+    const float& x5 = mdsInGPU.anchorX[fifthMDIndex]/100;
+    const float& y1 = mdsInGPU.anchorY[firstMDIndex]/100;
+    const float& y2 = mdsInGPU.anchorY[secondMDIndex]/100;
+    const float& y3 = mdsInGPU.anchorY[thirdMDIndex]/100;
+    const float& y4 = mdsInGPU.anchorY[fourthMDIndex]/100;
+    const float& y5 = mdsInGPU.anchorY[fifthMDIndex]/100;
+
+    float residual = 0;
+    float error = 0;
+    float x_center=g/100, y_center=f/100; 
+    float x_init=mdsInGPU.anchorX[thirdMDIndex]/100;
+    float y_init=mdsInGPU.anchorY[thirdMDIndex]/100;
+    float z_init=mdsInGPU.anchorZ[thirdMDIndex]/100;
+    float rt_init=mdsInGPU.anchorRt[thirdMDIndex]/100; //use the second MD as initial point
+
+    if (moduleType3==1)  // 1: if MD3 is in 2s layer
     {
-        slope = (z2 -z1)/(rt2 - rt1);
+        x_init=mdsInGPU.anchorX[secondMDIndex]/100;
+        y_init=mdsInGPU.anchorY[secondMDIndex]/100;
+        z_init=mdsInGPU.anchorZ[secondMDIndex]/100;
+        rt_init=mdsInGPU.anchorRt[secondMDIndex]/100;
     }
-    else
+
+    //start from a circle of inner T3.
+    // to determine the charge
+    int charge=0;
+    float slope3c=(y3-y_center)/(x3-x_center);
+    float slope1c=(y1-y_center)/(x1-x_center);
+    if((y3-y_center)>0 && (y1-y_center)>0) 
+    {
+        if (slope3c>slope1c) charge=-1; 
+        else if (slope3c<slope1c) charge=1;
+        if (slope1c>0 && slope3c<0) charge=-1;
+        if (slope1c<0 && slope3c>0) charge=1;
```

Comment:
Yes, sorry for the confusion. This is indeed set twice, because the latter 2 "if"s are the exclusive cases and set it hard by hand. That means usually if slope3c<slope1c, charge=1. But if slope1c>0 && slope3c<0, always set charge =-1. 
To be more clear, L678 and L679 are the general cases for 90% of the events. But if L680 and L681 conditions are satisfied, the charge is incorrect, so I fixed manually.

Reply:
Thanks for the explanation. However, can you make the code a bit more logical, so that the cases where two conditions are fulfilled are better separated? If I understand the logic correctly, it would make more sense if you had written:
```suggestion
        if (slope1c>0 && slope3c<0) charge=-1;
        else if (slope3c<slope1c) charge=1; # Write a helpful comment here to explain
        else if (slope1c<0 && slope3c>0) charge=1;
        else if (slope3c>slope1c) charge=-1; # Write a helpful comment here to explain
```
Of course, please make sure that this logic works as you want it to, because maybe I have missed some of the details.

Reply:
Thank you for reordering this! I've made the new conditions as suggested and also add comments about this part of the charge definition. Hope that helps with people understanding.